Add Reassure mobile performance testing library

derekblank opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Adds the [Reassure](https://github.com/callstack/reassure) performance testing library to test mobile performance on React Native components.

## Why?
To help analyze and address performance issues within mobile components. 

## How?
In addition to adding the Reassure library, the PR also adds mobile performance tests.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

derekblank committed:
Add sample reassure perf-test

derekblank committed:
Add reassure library

derekblank committed:
Add text input performance tests

dcalhoun committed:
test: Fix Reassure test configuration

Only 1 of the 2 tests works. I left notes as to why the other fails.
Also, I quickly used the `toBeTruthy` assertion, it may not be the most
appropriate. The `toBeInTheDocument` assertion is only relevant to DOM
tests, though. I do not believe we can use that.

derekblank committed:
Add RichText performance test

derekblank committed:
Update package-lock.json and Reassure current.perf hashes

dcalhoun commented on .reassure/current.perf:
```diff
@@ -0,0 +1 @@
+{"metadata":{"branch":"rnmobile/add-reassure","commitHash":"e4beb96b79fa5ac90ac12f6a4d3c81f8d6a10489"}}
```
Do you know why this file exists both here and `packages/react-native-editor/.reassure/current.perf`? Should we have one or two copies of this file?

derekblank replied:
Based on the [Reassure docs](https://github.com/callstack/reassure#measuring-test-performance), these current.perf files are generated when `yarn reassure` is ran (or in our case, `npm run native test:perf`), and they do not need to be committed to version control, so I've removed them from git control and added them to `.gitignore`. I am not exactly sure what these files are meant to do based on the documentation, but it seems like they may map a commit hash to a specific set of performance results. Let me know if you interpret this documentation differently and think they need to be included, however. 

> Should we have one or two copies of this file?

If we do decide to include these, at the very least, I do not think we need two copies of this file.

dcalhoun replied:
Agreed. Based on the documentation, it does make sense to not commit these to version control. It appears to be internal tooling for the CLI script. 

My best guess as to why we two copies were created is that we may have run the `reassure` script from different directories while exploring setup initially.

dcalhoun commented on package.json:
```diff
@@ -221,6 +221,7 @@
 		"react-native-url-polyfill": "1.1.2",
 		"react-refresh": "0.10.0",
 		"react-test-renderer": "18.2.0",
+		"reassure": "0.7.1",
```
When I run `npm install` I see the following diff. Will you try `npm run distclean && npm install` to see if you see any changes in your `package-lock.json`?

```diff
diff --git a/package-lock.json b/package-lock.json
index 7151bff2df..fa549eab19 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -29769,7 +29769,7 @@
 		"co": {
 			"version": "4.6.0",
 			"resolved": "https://registry.npmjs.org/co/-/co-4.6.0.tgz",
-			"integrity": "sha1-bqa989hTrlTMuOR7+gvz+QMfsYQ=",
+			"integrity": "sha512-QVb0dM5HvG+uaxitm8wONl7jltx8dqhfU33DcqtOZcLSVIKSDDLDi7+0LbAKiyI8hD9u42m2YxXSkMGWThaecQ==",
 			"dev": true
 		},
 		"code-point-at": {
@@ -43881,7 +43881,7 @@
 		"lodash.isequal": {
 			"version": "4.5.0",
 			"resolved": "https://registry.npmjs.org/lodash.isequal/-/lodash.isequal-4.5.0.tgz",
-			"integrity": "sha1-QVxEePK8wwEgwizhDtMib30+GOA="
+			"integrity": "sha512-pDo3lu8Jhfjqls6GkMgpahsF9kCyayhgykjyLMNFTKWrpVdAQtYyB4muAMWozBB4ig/dtWAmsMxLEI8wuz+DYQ=="
 		},
 		"lodash.ismatch": {
 			"version": "4.4.0",

```

derekblank replied:
After resetting npm and running `npm run distclean && npm install`, I still was not able to produce this diff on node v14.21.2 (npm v6.14.18). It seems like [sha512](https://github.com/npm/npm/issues/16938) is the SHA hash we should use, however. Due to my npm version issue, sha1 may have been used when I ran the original package-lock.json file changes. So, perhaps we should commit the diff that you are able to produce?

dcalhoun replied:
Here is my current environment: 

```zsh
❯ node -v
v14.20.0

❯ npm -v
6.14.18
```

After upgrading Node.js and `npm` to match your environment, I no longer see lock file changes when running `npm run distclean && npm install`. So, I believe we are good to keep the changes as-is.

dcalhoun commented on packages/block-editor/src/components/rich-text/test/index.perf-test.js:
```diff
@@ -0,0 +1,45 @@
+/**
+ * External dependencies
+ */
+import { measurePerformance } from 'reassure';
```
Does it make sense to export this from `test/helpers`? I believe we export `@testing-library/react-native` from there to reduce the references to the third-party package in the hopes of making refactors easier.

derekblank replied:
Updated.

dcalhoun commented on packages/block-editor/src/components/rich-text/test/index.perf-test.js:
```diff
@@ -0,0 +1,45 @@
+ * External dependencies
+ */
+import { measurePerformance } from 'reassure';
+import { fireEvent } from '@testing-library/react-native';
```
Are we able to leverage `test/helpers` here and the rest of the test files? 
```suggestion
import { fireEvent } from 'test/helpers';
```

derekblank replied:
Updated.

dcalhoun commented on packages/block-editor/src/components/rich-text/test/index.perf-test.js:
```diff
@@ -0,0 +1,45 @@
+import { RichText } from '@wordpress/block-editor';
+
+describe( 'RichText Performance', () => {
+	it( 'measures performance', async () => {
```
The current description is a bit repetitive and uninformative. What do you think about the following approach for this and the other tests? 
```suggestion
describe( 'RichText', () => {
	it( 'performance is stable', async () => {
```

derekblank replied:
Updated.

dcalhoun commented on packages/block-editor/src/components/rich-text/test/index.perf-test.js:
```diff
@@ -0,0 +1,45 @@
+		const scenario = async () => {
+			const richTextInput = <RichText />;
+			// Simulate user typing text.
+			fireEvent( richTextInput, 'focus' );
```
This test fails currently, but that is expected to me as this line is improper to my understanding. This appears to set `richTextInput` to a separate copy of `RichText` than the one passed to `measurePerformance`. The test currently fails with the following error:

```
TypeError: Cannot read property 'parent' of undefined
```

My thought is that maybe this should reference `screen` passed to the `scenario` callback. 

```suggestion
		const scenario = async (screen) => {
			// Simulate user typing text.
			fireEvent( screen, 'focus' );
```

dcalhoun commented on packages/block-editor/src/components/rich-text/test/index.perf-test.js:
```diff
@@ -0,0 +1,45 @@
+
+		const changeTextOfRichText = ( richText, text ) => {
+			fireEvent( richText, 'focus' );
+			fireEvent( richText, 'onChange', {
+				nativeEvent: {
+					eventCount: 1,
+					target: undefined,
+					text,
+				},
+			} );
```
This fires a single event of `onChange`, which makes total sense for test preparation. Speaking at a higher level, while a user this event fires several times for each character. This raises two questions for me:

1. Does it make sense to mimic typing by triggering `onChange` for each character of a text set or is that fruitless? 
2. Is rendering and testing `RichText` alone a meaningful performance test or do we need a deeper editor tree? 

For the first question, it could be moot. If 1 `onChange` invocations equals 1 render, then 20 `onChange` invocations, in theory, equals 20 renders. If they are equal, is there value in mimicking typing via multiple events? 

For the second question, this likely comes down to _where performance issues originate_. Is the problem that `RichText` renders to frequently due to internal state? Is it due to props from parents? Is it due to context changes? It is difficult to know the most appropriate way to test performance without knowing the origin issue we want to avoid. Avoiding (and tracking) unnecessary re-renders is a good practice, but to what end? 

I share this not to say this test is incorrect, but to begin a discussion as to where/how performance tests provide meaningful value.

derekblank replied:
Good questions. 

> Does it make sense to mimic typing by triggering onChange for each character of a text set or is that fruitless?

I don't think it would be a bad thing to measure microinteractions like this. They may produce useful metrics, and the only negative I can see is spending too much time writing Reassure tests. In the beginning of this project, I  envisioned that the Reassure test suite would need to expand to potentially every component in order to measure performance effectively. I'm re-evaulating that notion as we continue to write tests and explore testing organization. For the purposes of this PR, I wasn't intending for this to be a very robust test case, but I would very much welcome further discussion of how the structure of these tests could evolve into a suite of robust test cases.

> Is rendering and testing RichText alone a meaningful performance test or do we need a deeper editor tree?

This was very much on my mind while writing the test as well -- I was trying to think of ways to avoid false positives or false reinforcement within the test structure itself. Having done a small amount of performance testing with Reassure, I think considering the depth of the editor tree is a very good topic to explore (or potentially discuss with the creators of the library) to see if the depth of nodes makes an impact on performance results, particularly as it relates to tracing exactly where re-renders originate.

dcalhoun commented on packages/block-editor/src/components/rich-text/test/index.perf-test.js:
```diff
@@ -0,0 +1,45 @@
+				'<strong>Bold</strong> <em>italic</em> <s>strikethrough</s> text'
+			);
+
+			// Check if the onChange is called and the state is updated.
+			expect( onChange ).toHaveBeenCalledTimes( 1 );
```
I am unable to interpret the specific value this assertion provides. Is this in place merely to appease the linter warning or is asserting the callback testing something specific?

dcalhoun commented on packages/components/src/mobile/html-text-input/test/index.perf-test.js:
```diff
@@ -0,0 +1,42 @@
+			const htmlTextInput = findContentTextInput( screen );
+			fireEvent( htmlTextInput, 'changeText', 'text' );
+
+			// Check if the onChange is called and the state is updated.
+			expect( onChange ).toHaveBeenCalledTimes( 1 );
+			expect( onChange ).toHaveBeenCalledWith( 'text' );
```
Similar to an earlier comment, these assertions seem to test internals that may not be necessary to test. The next assertion targeting the display value seems to be more appropriate and provide similar coverage. WDYT?

derekblank replied:
These assertions initially started as linter requirements. Although they don't affect the outcome of the performance results, I think they could be expanded here to help describe what the test is measuring.

dcalhoun replied:
> Although they don't affect the outcome of the performance results, I think they could be expanded here to help describe what the test is measuring.

I suppose given we are testing this single unit in isolation, asserting the public API (callbacks and props) functions as expected makes sense. 👍🏻

dcalhoun reviewed (commented):
This is great progress and exploration! Thank you for doing this work. 

I left comments on the approach in a few places, and raised some larger questions in regards to what we want to test exactly. I recognize the latter is a bit more open-ended and difficult to answer.

derekblank committed:
Export measurePerformance from test helpers and update performance test headers

derekblank committed:
Remove HTMLTextInput performance test and update current perf

derekblank committed:
Remove Reassure current.perf files from git control

dcalhoun committed:
test: Improve mobile editor performance tests (#48101)

* test: Linter considers measurePerfomrance an assertion

Performance tests often only assert the performance of a component via
`measurePerformance`. This change adds `measurePerformance` to the list
of functions that are considered assertions to avoid erroneous lint
warnings.

* test: Fix errors in Rich Text performance test

- Remove unnecessary lint warning disable line caused by erroneous
  placement of `screen` argument.
- Query UI components via user-facing attributes, e.g. a11y label.
- Pass mock function to required `onChange` prop to `RichText`
  component.

* test: Rename native editor performance test files

To avoid web test scripts from including the performance files for the
native mobile editor, we must include the term "native" in the file
name. This renames the test files to follow the existing pattern of not
including "test" in the file name.

derekblank committed:
Merge trunk and resolve conflict in packages/react-native-editor/package.json

derekblank committed:
Update package-lock.json

derekblank committed:
Add Reassure CI test script and update Github performance workflow

derekblank committed:
Remove Gallery block perf-native test

derekblank committed:
Merge remote-tracking branch 'origin' into rnmobile/add-reassure

derekblank commented:
Closing in favor of https://github.com/WordPress/gutenberg/pull/49221 due to merge conflicts:
* https://github.com/WordPress/gutenberg/pull/49221
